sydent/util/emailutils.py

```diff
@@ -73,7 +74,14 @@ def sendEmail(
     # conflicting with it.
     allSubstitutions["multipart_boundary"] = generateAlphanumericTokenOfLength(32)
 
-    mailString = open(templateFile).read() % allSubstitutions
+    # Render template with Jinja if using Jinja template
+    if templateFile.endswith(".j2"):
+        with open(templateFile) as file_:
+            template = Template(file_.read())
```

Comment:
```suggestion
        with open(templateFile) as template_file:
            template = Template(template_file.read())
```

but probably also factor this out so it just reads the file and then does either thing with the contents (only one `open` call); put the `if` block in the `with` block?

Reply:
I wonder if we should be using an environment for Jinja instead of creating templates manually. (I think you get caching and such for free if you do that.)

Also we need to consider autoescaping here. I *think* some of the substitutions come from user's so could be used to inject HTML or other things into the email templates. Doing `autoescape=select_autoescape()` *might* do what you want, but I'm not sure.

Also it would be ideal if we had tests for this. 😄

Reply:
I think creating an environment for Jinja is indeed the way to go. Loading a template takes a little bit of time (apparently it's converted into Python code and compiled), so even though we don't appear to have bothered for old-school templates, we probably should for Jinja templates.

Escaping is a bit of a nuisance here: old-school templates don't auto-escape, so the current code 'should' be doing that already (I saw a few indicators that it at least tries).
Using auto-escaping is something that I'm a big fan of too though.
However, you don't want to escape in the plaintext part of the e-mail -- you can opt out of autoescaping by writing e.g. `{{ room_name | safe }}`.

I think the way forward may be to take care of doing all the auto-escaping for the old-school templates -- just escape ALL the fields. (I reckon we already show escaped HTML in the plaintext fragments currently, so it won't be the end of the world to carry on).


One idea, which @clokep can comment on:
You could have a function e.g. `def load_template(self, template_name: str) -> Callable[dict, str]:` which accepts a template name and returns a function that: accepts a dict of parameters and returns the templated result.
You can store these functions in a dictionary (template name -> function) and then avoid loading them every single time we need to use the template.

(Alternatively, if the Jinja2 environment does the caching for free, then we could potentially just go with that. Just trying to think about whether we can get away from having to test whether a template is old or J2 each time we wish to use it -- my function cache would get around that, but will defer to Patrick on what he thinks is best)

Reply:
I think what you want to do is create the template object once and re-use it (instead of re-creating it each time). This could probably be done in a follow-up PR though if it is complex / requires rearchitecting.

A good way to structure the code to handle the escaping might be:

```
If the template is Jinja:
    Create a template and enable auto-escape. (Also probably disable it for the plaintext part / headers.)
    Use the provided substitutions directly.
Else:
    Manually escape the data for HTML / URLs.
    Do the formatting that we do now for the plaintext templates.
```

This should avoid weird double-escaping, but still let Jinja do it's own escaping. (Note that Jinja also has some helpers for *URL* escaping which we might want to use...)

Reply:
Okay I am looking at creating an environment now (which does cache, which I assume takes care of the issue around re-creating?). It does seem like it might require a little re-wiring, although maybe not too complex? If it gets too complex I will open up a follow-up PR.

Reply:
Also just to make sure I've got this all clarified and summarised, what needs to be done is:
-use an environment/figure out how to not re-create templates every time
-figure out escaping for both the non-jinja templates and the new jinja templates
-add some tests to show that jinja is rendering our info correctly and escaping is working
Let me know if anything is incorrect or I'm missing something!

Reply:
That sounds about right from what I was thinking!